### PR TITLE
[SPARK-14684] [SQL] Verification of Partition Specs in SessionCatalog

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

When users inputting invalid partition spec, we might not be able to catch and issue the error messages. Sometimes, it could cause a disaster result. For example, before the PR, when we alter a table and drop a partition with using unknown partitioning columns, we drop all the partitions due to a bug/defect in Hive Metastore API. 

This PR is to add all the partition-related verification in `SessionCatalog`.
#### How was this patch tested?

Fixed the illegal existing test cases and added new test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14 @cloud-fan Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 1]
```diff
@@ -368,6 +368,12 @@ class SessionCatalog(
       tableName: TableIdentifier,
       parts: Seq[CatalogTablePartition],
       ignoreIfExists: Boolean): Unit = {
+    val tableMetadata = getTableMetadata(tableName)
+    // Verify if each part exactly matches the partition spec of the table definition
+    parts.foreach { p =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: So the order of partition column names in give partition spec is strict? I.e., given partition specs is ["b", "a"] and the table metadata is ["a", "b"]. We don't allow this?

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: We do not allow it. Here, `requireExactMatchedPartitionSpec ` is to verify if two sequences are exactly identical.

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Hive allows it. In RDBMS, the order of partitioning columns matters. In Hive partitioning, each partition only has a static partitioning value. Anyway, let me follow what Hive did. Thanks!

Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Now, different orders are allowed.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14

**Comment by gatorsmile** [comment]
cc @cloud-fan : )

**Comment by cloud-fan** [comment]
I may not have enough energy to review this today, cc @andrewor14 do you have time to look at it first?

**Comment by gatorsmile** [comment]
This is not urgent at all. Just fixing the bugs. Please ignore my ping. 

Have a good day! : )

**Comment by gatorsmile** [comment]
This is closed. It will be part of another PR: https://github.com/apache/spark/pull/12801

**Pull request closed without merging**